// @base-ui/react/esm/menu/popup/MenuPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let MenuPopupDataAttributes = function (MenuPopupDataAttributes) {
  /**
   * Present when the menu is open.
   */
  MenuPopupDataAttributes[MenuPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the menu is closed.
   */
  MenuPopupDataAttributes[MenuPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the menu is animating in.
   */
  MenuPopupDataAttributes[MenuPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the menu is animating out.
   */
  MenuPopupDataAttributes[MenuPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  MenuPopupDataAttributes["side"] = "data-side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  MenuPopupDataAttributes["align"] = "data-align";
  /**
   * Present if animations should be instant.
   * @type {'click' | 'dismiss' | 'group'}
   */
  MenuPopupDataAttributes["instant"] = "data-instant";
  return MenuPopupDataAttributes;
}({});
