// @base-ui/react/esm/menu/arrow/MenuArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let MenuArrowDataAttributes = function (MenuArrowDataAttributes) {
  /**
   * Present when the menu popup is open.
   */
  MenuArrowDataAttributes[MenuArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the menu popup is closed.
   */
  MenuArrowDataAttributes[MenuArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  MenuArrowDataAttributes["side"] = "data-side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  MenuArrowDataAttributes["align"] = "data-align";
  /**
   * Present when the menu arrow is uncentered.
   */
  MenuArrowDataAttributes["uncentered"] = "data-uncentered";
  return MenuArrowDataAttributes;
}({});
